src/config.cpp

```diff
@@ -201,6 +202,172 @@ std::string toAbsolute(const fs::path& base, const fs::path& path) {
+    const auto element = it.find(name);
+    if (element != it.end()) {
+        buf = element->get<Type>();
+        raiseIfInvalidEnum(name, buf, element->dump(), std::is_enum<Type>());
+    } else if (default_value != nonstd::nullopt) {
+        buf = default_value.value();
+    }
+}
+
+SimulationConfig::Input parseInputModule(const nlohmann::json& valueIt,
+                                         const SimulationConfig::InputBase::Module module,
+                                         const std::string& basePath,
+                                         int randomSeed,
+                                         const std::string& debugStr) {
+    using Module = SimulationConfig::InputBase::Module;
+
+    const auto moduledebugStr = fmt::format("Unknown module for the input_type in {}",
+                                            debugStr);
+    const auto parseCommon = [&](auto& input) {
+        input.module = module;
+        parseMandatory(valueIt, "input_type", debugStr, input.inputType);
+        parseMandatory(valueIt, "delay", debugStr, input.delay);
+        parseMandatory(valueIt, "duration", debugStr, input.duration);
+        parseMandatory(valueIt, "node_set", debugStr, input.nodeSet);
+    };
+
+    switch (module) {
+    case Module::linear: {
+        SimulationConfig::InputLinear ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "amp_start", debugStr, ret.ampStart);
+        parseOptional<double>(valueIt, "amp_end", ret.ampEnd, ret.ampStart);
+        return ret;
+    }
+    case Module::relative_linear: {
+        SimulationConfig::InputRelativeLinear ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "percent_start", debugStr, ret.percentStart);
+        parseOptional<double>(valueIt, "percent_end", ret.percentEnd, ret.percentStart);
+        return ret;
+    }
+    case Module::pulse: {
+        SimulationConfig::InputPulse ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "amp_start", debugStr, ret.ampStart);
+        parseMandatory(valueIt, "width", debugStr, ret.width);
+        parseMandatory(valueIt, "frequency", debugStr, ret.frequency);
+        parseOptional<double>(valueIt, "amp_end", ret.ampEnd, ret.ampStart);
+        return ret;
+    }
+    case Module::subthreshold: {
+        SimulationConfig::InputSubthreshold ret;
+        parseCommon(ret);
+        parseMandatory(valueIt, "percent_less", debugStr, ret.percentLess);
+        return ret;
+    }
+    case Module::noise: {
+        SimulationConfig::InputNoise ret;
+        parseCommon(ret);
+        const auto mean = valueIt.find("mean");
+        const auto mean_percent = valueIt.find("mean_percent");
+
+        if (mean != valueIt.end()) {
+            parseOptional(valueIt, "mean", ret.mean);
+        }
+
+        if (mean_percent != valueIt.end()) {
+            parseOptional(valueIt, "mean_percent", ret.meanPercent);
+        }
+
+        if (ret.mean.has_value() && ret.meanPercent.has_value()) {
```

Comment:
Shall we combine the two exceptions? I think it is fine to throw as long as `ret.mean.has_value() == ret.meanPercent.has_value()`

Reply:
I think this makes it more clear, so I'm going to keep it this way.